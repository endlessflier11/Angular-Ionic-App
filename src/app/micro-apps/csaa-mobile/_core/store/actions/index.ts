export * from './config.actions';
export * from './metadata.actions';
export * from './service-locator.actions';
export * from './auth.actions';
export * from './fetch.actions';
export * from './policy.actions';
export * from './customer.actions';
export * from './payment.actions';
export * from './claim.actions';
export * from './user-setting.actions';
