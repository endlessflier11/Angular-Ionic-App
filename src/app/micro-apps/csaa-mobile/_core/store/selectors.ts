import { FetchStatus } from '../interfaces';
import {
  AuthStateModel,
  CONFIG_STATE_TOKEN,
  ConfigStateModel,
  CSAA_APP_STATE_TOKEN,
  CSAA_AUTH_STATE_TOKEN,
  CUSTOMER_STATE_TOKEN,
  CustomerStateModel,
} from './states/state.interfaces';

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace ConfigSelector {
  export const state = (appState: { csaa_app: { csaa_config: ConfigStateModel } }) =>
    appState[CSAA_APP_STATE_TOKEN][CONFIG_STATE_TOKEN];

  export const configAndEndpointsLoaded = (appState: {
    csaa_app: { csaa_config: ConfigStateModel };
  }): boolean => {
    const configState = ConfigSelector.state(appState);
    return configState?.configLoaded && configState?.endpointsLoadState === FetchStatus.SUCCESS;
  };

  export const storeInitialized = (appState: {
    csaa_app: { csaa_config: ConfigStateModel };
  }): boolean => {
    const configState = ConfigSelector.state(appState);
    return configState?.storeInitialized;
  };
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace AuthSelector {
  export const state = (appState: { csaa_app: { csaa_auth: AuthStateModel } }) =>
    appState[CSAA_APP_STATE_TOKEN][CSAA_AUTH_STATE_TOKEN];
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace CustomerSelector {
  export const state = (appState: { csaa_app: { csaa_customer: CustomerStateModel } }) =>
    appState[CSAA_APP_STATE_TOKEN][CUSTOMER_STATE_TOKEN];
}
