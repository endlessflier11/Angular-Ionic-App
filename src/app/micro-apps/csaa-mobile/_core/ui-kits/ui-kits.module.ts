import { NgModule } from '@angular/core';
import { CommonModule, DatePipe } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { IonicModule } from '@ionic/angular';
import { ThemeIconComponent } from './theme-icon/theme-icon.component';
import { WorkingHoursComponent } from './working-hours/working-hours.component';
import { MyPolicyCardComponent } from './my-policy-card/my-policy-card.component';
import { SkeletonCardComponent } from './skeleton-card/skeleton-card.component';
import { BackButtonComponent } from './back-button/back-button.component';
import { PhonePipe } from './pipes/phone-number.pipe';
import { CsaaAutoIdCardComponent } from './csaa-auto-id-card/csaa-auto-id-card.component';
import { CsaaLoadingComponent } from './csaa-loading/csaa-loading.component';
import { ContentRendererComponent } from './content-renderer/content-renderer.component';
import { NationalCustomAlertModalComponent } from './national-custom-alert/national-custom-alert-modal';
import { SafeDataPipe } from './pipes/safe-data/safe-data.pipe';
import { DocumentActionModalComponent } from './document-action-modal/document-action-modal.component';
import { CsaaPolicySubtitlesComponent } from './csaa-policy-subtitles/csaa-policy-subtitles.component';
import { PaperlessTermsComponent } from './paperless-terms/paperless-terms.component';
import { PaperlessTermsModalComponent } from './paperless-terms-modal/paperless-terms-modal.component';
import { AppleWalletButtonComponent } from './apple-wallet-button/apple-wallet-button.component';

const UI_COMPONENTS = [
  ThemeIconComponent,
  WorkingHoursComponent,
  MyPolicyCardComponent,
  SkeletonCardComponent,
  BackButtonComponent,
  CsaaAutoIdCardComponent,
  CsaaLoadingComponent,
  ContentRendererComponent,
  NationalCustomAlertModalComponent,
  DocumentActionModalComponent,
  CsaaPolicySubtitlesComponent,
  PaperlessTermsComponent,
  PaperlessTermsModalComponent,
  AppleWalletButtonComponent,
];

const UI_PIPES = [PhonePipe, SafeDataPipe];

@NgModule({
  imports: [CommonModule, ReactiveFormsModule, IonicModule],
  declarations: [...UI_COMPONENTS, ...UI_PIPES],
  exports: [...UI_COMPONENTS, ...UI_PIPES],
  providers: [DatePipe],
})
export class UiKitsModule {}
