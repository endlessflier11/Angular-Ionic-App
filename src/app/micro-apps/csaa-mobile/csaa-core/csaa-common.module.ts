import { ModuleWithProviders, NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FileOpener } from '@ionic-native/file-opener/ngx';
import { File } from '@ionic-native/file/ngx';
import { CardIO } from '@ionic-native/card-io/ngx';
import { SecureStorageEcho } from '@ionic-native/secure-storage-echo/ngx';

@NgModule({
  imports: [CommonModule],
  providers: [FileOpener, File, CardIO, SecureStorageEcho],
})
export class CsaaCommonModule {
  constructor(@Optional() @SkipSelf() parentModule?: CsaaCommonModule) {
    if (parentModule) {
      throw new Error('CsaaCommonModule is already loaded. Import it in the CsaaCoreModule only');
    }
  }

  static forRoot(): ModuleWithProviders<CsaaCommonModule> {
    return {
      ngModule: CsaaCommonModule,
    };
  }
}
