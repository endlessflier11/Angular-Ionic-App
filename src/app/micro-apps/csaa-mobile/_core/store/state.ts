import { Injectable } from '@angular/core';
import { State } from '@ngxs/store';
import { ConfigState } from './states/config.state';
import { MetadataState } from './states/metadata.state';
import { CsaaAuthState } from './states/auth.state';
import { ContactInfoState } from './states/contact-info.state';
import { FetchState } from './states/fetch.state';
import { CUSTOMER_CHILD_STATES, CustomerState } from './states/customer.state';
import { CSAA_APP_STATE_TOKEN } from './states/state.interfaces';
import { UserSettingsState } from './states/user-setting.state';

/*
 * - __csaa_app
 * ├── config
 * ├── user_setting
 * ├── metadata
 * ├── contact // TODO: consume state withing the app
 * ├── fetch
 * ├── auth
 * └── customer
 *     ├── policies
 *     ├── payments
 *     └── claims
 * */
const CHILD_STATES = [
  ConfigState,
  MetadataState,
  ContactInfoState,
  CsaaAuthState,
  FetchState,
  CustomerState,
  UserSettingsState,
];

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface CsaaAppStateModel {}

@State<CsaaAppStateModel>({
  name: CSAA_APP_STATE_TOKEN,
  children: CHILD_STATES,
})
@Injectable()
export class CsaaAppState {}

export const CSAA_STORES = [CsaaAppState, ...CHILD_STATES, ...CUSTOMER_CHILD_STATES];
