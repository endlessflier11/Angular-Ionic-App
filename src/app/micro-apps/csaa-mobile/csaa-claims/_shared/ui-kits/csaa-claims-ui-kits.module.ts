import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { CommonModule } from '@angular/common';
import { UiKitsModule } from '../../../_core/ui-kits/ui-kits.module';
import { AgentCardComponent } from './agent-card/agent-card.component';
import { ClaimsHistoryCardComponent } from './claims-history-card/claims-history-card.component';
import { ClaimsInfoCardComponent } from './claims-info-card/claims-info-card.component';
import { WhatToDoCardComponent } from './what-to-do-card/what-to-do-card.component';

const SHARED_COMPONENTS = [
  AgentCardComponent,
  ClaimsHistoryCardComponent,
  ClaimsInfoCardComponent,
  WhatToDoCardComponent,
];

@NgModule({
  declarations: [...SHARED_COMPONENTS],
  imports: [CommonModule, IonicModule, UiKitsModule],
  exports: [...SHARED_COMPONENTS],
})
export class CsaaClaimsUiKitsModule {}
