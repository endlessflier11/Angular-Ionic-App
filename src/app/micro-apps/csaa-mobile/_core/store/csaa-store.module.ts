import { NgModule } from '@angular/core';
import { environment } from 'src/environments/environment';
import { NgxsModule } from '@ngxs/store';
import { CSAA_STORES } from './state';

@NgModule({
  imports: [
    NgxsModule.forRoot(CSAA_STORES, {
      developmentMode: !environment.production,
    }),
    NgxsModule.forFeature(CSAA_STORES),
  ],
})
export class CsaaStoreModule {}
