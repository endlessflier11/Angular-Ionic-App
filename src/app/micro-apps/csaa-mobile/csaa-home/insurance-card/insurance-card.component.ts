import { Component, Input, OnDestroy } from '@angular/core';
import { noop } from '../../_core/helpers';
import { Policy, PolicyType } from '../../_core/interfaces/policy.interface';
import { AnalyticsService, RouterService } from '../../_core/services';
import { Category, EventName, EventType } from '../../_core/interfaces';
import { PolicyHelper } from '../../_core/shared/policy.helper';
import { PolicyDocumentHelper } from '../../_core/shared/policy-document.helper';
import { DocumentType } from '../../_core/interfaces/document.interface';
import { Select } from '@ngxs/store';
import { FetchState } from '../../_core/store/states/fetch.state';
import { PolicyAction } from '../../_core/store/actions';
import { combineLatest, Observable } from 'rxjs';
import { SubSink } from '../../_core/shared/subscription.helper';

@Component({
  selector: 'csaa-insurance-card',
  templateUrl: './insurance-card.component.html',
  styleUrls: ['./insurance-card.component.scss'],
})
export class InsuranceCardComponent implements OnDestroy {
  @Input() policies: Policy[] = [];
  @Select(FetchState.needsFetching(PolicyAction.LoadPolicies))
  policiesNeedsFetching$: Observable<boolean>;
  @Select(FetchState.isLoading(PolicyAction.LoadPolicies)) loadingPolicies$: Observable<boolean>;

  private readonly subsink = new SubSink();
  // eslint-disable-next-line @typescript-eslint/naming-convention
  public PolicyType = PolicyType;
  public loading: boolean;

  constructor(
    private routerService: RouterService,
    private analyticsService: AnalyticsService,
    private policyDocumentHelper: PolicyDocumentHelper
  ) {
    this.subsink.add(
      combineLatest([this.loadingPolicies$, this.policiesNeedsFetching$]).subscribe((values) => {
        this.loading = values.find(Boolean);
      })
    );
  }

  ngOnDestroy() {
    this.subsink.unsubscribe();
  }

  async insuranceCardClickHandler(policy: Policy) {
    if (policy.type === PolicyType.Auto) {
      this.analyticsService.trackEvent(EventName.AUTO_ID_CARDS_ACCESSED, Category.documents, {
        event_type: EventType.LINK_ACCESSED,
        link: 'Proof Of Insurance',
        ...AnalyticsService.mapPolicy(policy),
      });

      this.routerService
        .navigateForward('csaa.poi.index', { policyNumber: policy.number })
        .then(noop);
    } else if (policy.type === PolicyType.Home) {
      this.saveIdCardsToDevice(policy);
    }
  }

  // we don't show pup for Proof Of insurance cards
  get filteredPolicies() {
    return this.policies ? this.policies.filter((p) => p.type !== PolicyType.PUP) : [];
  }

  private saveIdCardsToDevice(policy: Policy) {
    this.analyticsService.trackEvent(EventName.PROOF_OF_INSURANCE_VIEWED, Category.documents, {
      event_type: EventType.FILE_DOWNLOADED,
      link: 'Proof Of Insurance',
      ...AnalyticsService.mapPolicy(policy),
    });

    const productType = PolicyHelper.prodTypeFromPolicyType(policy.type);
    return this.policyDocumentHelper.openDocument(
      policy.number,
      productType,
      DocumentType.Declarations
    );
  }
}
