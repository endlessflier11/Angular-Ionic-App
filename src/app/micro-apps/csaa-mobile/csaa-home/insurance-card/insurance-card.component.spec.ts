import { ComponentFixture, TestBed } from '@angular/core/testing';
import { IonicModule, LoadingController, Platform } from '@ionic/angular';

import { PolicyService } from '../../_core/services/policy.service';
import { PolicyDocumentHelper } from '../../_core/shared/policy-document.helper';
import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';

import { InsuranceCardComponent } from './insurance-card.component';

import { CsaaPaymentsUiKitsModule } from '../../csaa-payments/_shared/ui-kits/csaa-payments-ui-kits.module';
import { CsaaCoreModule } from '../../csaa-core/csaa-core.module';
import { CsaaStoreModule } from '../../_core/store/csaa-store.module';

import { PageTestingModule, StoreTestBuilder } from '@app/testing';
import { Store } from '@ngxs/store';

describe('InsuranceCardComponent', () => {
  let component: InsuranceCardComponent;
  let fixture: ComponentFixture<InsuranceCardComponent>;

  beforeEach(async () => {
    try {
      await TestBed.configureTestingModule({
        declarations: [InsuranceCardComponent],
        providers: [
          {
            provide: LoadingController,
            useValue: { create: jest.fn(() => ({ present: jest.fn() })) },
          },
          {
            provide: PolicyService,
            useValue: {
              getInsuredFirstName: jest.fn(),
              refreshPolicies: jest.fn(),
              listPolicies: jest.fn(),
              cleanUpPoliciesSubscriptions: jest.fn(),
            },
          },
          {
            provide: PolicyDocumentHelper,
            useValue: {
              openDocument: jest.fn(),
            },
          },
        ],
        imports: [
          PageTestingModule.withConfig({
            providesStorage: true,
            providesRouter: true,
            providesAnalytics: true,
            providesAlert: true,
          }),
          IonicModule,
          UiKitsModule,
          CsaaPaymentsUiKitsModule,
          CsaaCoreModule.forRoot(),
          CsaaStoreModule,
        ],
      }).compileComponents();

      const store = TestBed.inject(Store);
      StoreTestBuilder.withDefaultMocks().resetStateOf(store);
      await TestBed.inject(Platform).ready();

      fixture = TestBed.createComponent(InsuranceCardComponent);
      component = fixture.componentInstance;
      fixture.detectChanges();
    } catch (error) {
      console.error(error);
      fail(error.message);
    }
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  // Todo: match snapshot for when policy is loaded
});
