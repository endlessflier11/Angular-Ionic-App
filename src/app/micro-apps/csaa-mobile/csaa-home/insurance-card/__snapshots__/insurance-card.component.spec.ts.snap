// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsuranceCardComponent should match snapshot 1`] = `
<csaa-insurance-card
  PolicyType={[Function Object]}
  __loadingPolicies$__selector={[Function Observable]}
  __ngContext__={[Function LRootView]}
  __policiesNeedsFetching$__selector={[Function Observable]}
  analyticsService={[Function AnalyticsMockService]}
  loading="undefined"
  policies={[Function Array]}
  policyDocumentHelper={[Function Object]}
  routerService={[Function RouterMockService]}
  subsink={[Function SubSink]}
>
  <ion-card
    class="csaa-card"
    data-test="Proof-of-Insurance"
  >
    <ion-card-header>
      <ion-card-title>
         Proof of Insurance 
      </ion-card-title>
    </ion-card-header>
    <ion-card-content
      class="csaa-item-list"
    >
      <ion-list>
      </ion-list>
    </ion-card-content>
  </ion-card>
</csaa-insurance-card>
`;
